### ISSUE TITLE
Revise test_qobj_headers_in_result_sims

### ISSUE THREAD

**Issue by diego-plan9** [Qiskit/qiskit-ibmq-provider#87]
### What is the expected behavior?

After #86, the `test_qobj_headers_in_result_sims` was marked as expected failure, as passing `object_header` to `assemble()` no longer preserves the header pristine:

```
custom_qobj_header = {'x': 1, 'y': [1, 2, 3], 'z': {'a': 4}}
qobj_header = QobjHeader.from_dict(custom_qobj_header)
qobj = assemble(circuits, backend=backend, qobj_header=qobj_header)
```

```
> print(qobj.header.as_dict())
{'x': 1, 'backend_name': 'ibmq_qasm_simulator', 'backend_version': '0.1.547', 'y': [1, 2, 3], 'z': {'a': 4}}
```

We should revise if this is by design and adjust the test or terra accordingly (for the test, simply reverting the order and updating `qobj.header` directly after `assemble()` would work). `test_qobj_headers_in_result_devices` might need similar measures.

**Comment by jwoehr** [comment]
I can adjust the tests accordingly, if desired. But is `assemble()` really doing what it should?
- Should the qobj dictionary be an ordered dict, and should `assemble()` likewise preserve and yield an ordered dict?
- Or should all consumers not presume any order to qobj's dict?

**Comment by diego-plan9** [comment]
It seems indeed be a (good) question better handled at terra - the order of the dicts should not be taken into account (as the specs don't define an order for the header), but `assemble()` is appending two extra keys to the original dict (which seems to be by design, but there is a small debate regarding it).

For the provider, if you are up for tackling this issue, can you go for modifying directly `qobj.header` _after_ `assemble()` (might be just reverting the order of the lines)? This would actually resemble better the original semantics of those two tests: they are actually testing whether the device returns the headers in `qobj.header`, regardless of how the `Qobj` is constructed.

**Comment by jwoehr** [comment]
Yes, I believe it should be that simple, changing order of operations. Looked at that when I first noticed the phenomenon in #53 but was not quite sure of the intent.

**Comment by jwoehr** [comment]
Hmm, I can't figure out how to get my fork in sync ... everything says they're the same, but when I check out master I don't see the `@expected_fail` annotation.

**Comment by diego-plan9** [comment]
Fixed via #94.
